### PR TITLE
fix: marketo bulk upload

### PR DESCRIPTION
**Pull request by utsabc**
## Description of the change

resolves INT-157

Fix for file read for marketo bulk upload

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

**Review by mihir-4116 — APPROVED** [review]

File: src/util/prometheus.js [hunk 1]
```diff
@@ -855,6 +867,49 @@ class Prometheus {
...
+        labelNames: [],
+      },
+      {
+        name: 'marketo_bulk_upload_fetch_job_create_response_time',
+        help: 'marketo_bulk_upload_fetch_job_create_response_time',
+        type: 'histogram',
+        labelNames: [],
+      },
+      {
+        name: 'marketo_bulk_upload_create_header_time',
```
Comment (utsabc) [inline src/util/prometheus.js:889]: There are some errors while we are starting transformer lets make sure there is no typo or anything in metrics

File: src/v0/destinations/marketo_bulk_upload/util.js [hunk 2]
```diff
@@ -1,98 +1,343 @@
...
+          500,
         );
       }
-      // Assuming none we should retry the remaining errors
-      throw new RetryableError(
-        resp.response.response.statusText || ACCESS_TOKEN_FETCH_ERR_MSG,
-        500,
+      return resp.response.access_token;
+    }
+    return null;
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/util.js:153]: Whats the behaviour of returning `null` in cache call back do we store null for the given key? If thats true we need to throw error here and not check null outside

File: src/v0/destinations/marketo_bulk_upload/util.js [hunk 3]
```diff
@@ -1,98 +1,343 @@
...
+ */
+const handleCommonErrorResponse = (resp, OpErrorMessage, OpActivity, config) => {
+  // checking for invalid/expired token errors and evicting cache in that case
+  // rudderJobMetadata contains some destination info which is being used to evict the cache
+  if (
+    authCache &&
+    resp.response?.errors &&
+    resp.response?.errors?.length > 0 &&
+    resp.response?.errors.some((errorObj) => errorObj.code === '601' || errorObj.code === '602')
+  ) {
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/util.js:76]: resp is bad name here please lets pick a non confusing name: maybe call `resp` to `apiCallResult`

File: src/v0/destinations/marketo_bulk_upload/config.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
+const ABORTABLE_CODES = ['ENOTFOUND', 'ECONNREFUSED', '601', '603', '605', '609', '610'];
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/config.js:1]: 'ENOTFOUND', 'ECONNREFUSED' we dont need these here, here we are only concered with Marketo application level error code or record level error code

File: src/v0/destinations/marketo_bulk_upload/util.js [hunk 5]
```diff
@@ -1,98 +1,343 @@
...
+    // for empty file the code is 1003 and that should be retried
+    stats.increment(OpActivity, {
+      status: 400,
+      state: 'Abortable',
+    });
+    throw new AbortedError(resp.response?.errors[0]?.message || OpErrorMessage, 400);
+  } else if (THROTTLED_CODES.includes(resp.response?.errors[0]?.code)) {
+    // for more than 10 concurrent uses the code is 615 and that should be retried
+    stats.increment(OpActivity, {
+      status: 500,
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/util.js:94]:
```suggestion
      status: 429,
```

File: src/v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 6]
```diff
@@ -344,39 +244,62 @@ const getImportID = async (input, config, fieldSchemaNames, accessToken) => {
...
+ * @param {*} input
+ * @param {*} config
+ * @returns returns the final response of fileUpload.js
+ */
 const responseHandler = async (input, config) => {
+  const accessToken = await getAccessToken(config);
+  // If token is null
+  if (!accessToken) {
+    throw new UnauthorizedError('Authorization failed');
+  }
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/fileUpload.js:258]: This should not be handled here

File: src/v0/destinations/marketo_bulk_upload/util.js [hunk 7]
```diff
@@ -333,11 +333,90 @@ const handleFileUploadResponse = (resp, successfulJobs, unsuccessfulJobs, reques
...
+  ) {
+    fieldArr =
+      fieldSchemaMapping.response.result && Array.isArray(fieldSchemaMapping.response.result)
+        ? fieldSchemaMapping.response.result[0]?.fields
+        : [];
+
+    fieldArr.forEach((field) => {
+      fieldMap[field?.name] = field?.dataType;
+    });
+  }
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/util.js:389]: Can this responsibility be offloaded to getFieldSchema fxn.

File: src/v0/destinations/marketo_bulk_upload/util.js [hunk 8]
```diff
@@ -333,11 +333,90 @@ const handleFileUploadResponse = (resp, successfulJobs, unsuccessfulJobs, reques
...
+    const { message, metadata } = event;
+    const { job_id } = metadata;
+
+    Object.entries(message).forEach(([paramName, paramValue]) => {
+      let expectedDataType = fieldMap[paramName];
+      const actualDataType = typeof paramValue;
+
+      // If expectedDataType is not one of the primitive data types, treat it as a string
+      if (!['string', 'number', 'boolean', 'undefined'].includes(expectedDataType)) {
+        expectedDataType = 'string';
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/util.js:402]: Can we comment some documentation from marketo. So essentially we are saying marketo doesnt support anything else than primitive data type?

File: src/v0/destinations/marketo_bulk_upload/fetchJobStatus.js [hunk 9]
```diff
@@ -97,38 +102,43 @@ const responseHandler = async (event, type) => {
...
+  const fieldSchemaMapping = await getFieldSchema(accessToken, config.munchkinId)
+  const unsuccessfulJobInfo = checkEventStatusViaSchemaMatching(event, fieldSchemaMapping)
 
-    for (const [key, val] of Object.entries(data)) {
-      // joining the parameter values sent from marketo match it with received data from server
-      if (val === `${elemArr.join()}`) {
-        // add job keys if warning/failure
-        if (!unsuccessfulJobIdsArr.includes(key)) {
-          unsuccessfulJobIdsArr.push(key);
+if (Object.keys(unsuccessfulJobInfo).length === 0 ){
```
Comment (utsabc) [inline src/v0/destinations/marketo_bulk_upload/fetchJobStatus.js:116]: Some jobs can be mismatched, some jobs can fail here when we check marketo response. Can't this be a scenario?

Comment (shrouti1507) [reply src/v0/destinations/marketo_bulk_upload/fetchJobStatus.js]: Logically yes. Updated it as if `(Object.keys(unsuccessfulJobInfo).length === 0 || type === 'fail' )`

Comment (shrouti1507) [reply src/v0/destinations/marketo_bulk_upload/fetchJobStatus.js]: still there will be some issue , will try to fix it

Comment (shrouti1507) [reply src/v0/destinations/marketo_bulk_upload/fetchJobStatus.js]: solved

File: src/v0/destinations/marketo_bulk_upload/poll.js [hunk 10]
```diff
@@ -18,106 +25,46 @@ const getPollStatus = async (event) => {
...
+  let statusCode = 500;
   let hasFailed;
-  let failedJobsURL;
-  let hasWarnings;
-  let warningJobsURL;
-  let errorResponse;
+  let HasWarning;
+  let error;
+  let InProgress = false;
+  const pollResp = await getPollStatus(event);
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply src/v0/destinations/marketo_bulk_upload/poll.js:55]: done

File: src/v0/destinations/marketo_bulk_upload/fileUpload.js (hunk 6 above)
Comment (shrouti1507) [reply src/v0/destinations/marketo_bulk_upload/fileUpload.js:242]: done

File: src/v0/destinations/marketo_bulk_upload/util.js (hunk 5 above)
Comment (shrouti1507) [reply src/v0/destinations/marketo_bulk_upload/util.js:94]: done

File: src/v0/destinations/marketo_bulk_upload/config.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+const RETRYABLE_CODES = ['EADDRINUSE', 'ECONNRESET', 'ETIMEDOUT', '713', '602', '604', '611'];
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply src/v0/destinations/marketo_bulk_upload/config.js:2]: done

File: src/v0/destinations/marketo_bulk_upload/config.js (hunk 4 above)
Comment (shrouti1507) [reply src/v0/destinations/marketo_bulk_upload/config.js:1]: done

File: src/v0/destinations/marketo_bulk_upload/util.js (hunk 3 above)
Comment (shrouti1507) [reply src/v0/destinations/marketo_bulk_upload/util.js:76]: done

File: src/v0/destinations/marketo_bulk_upload/util.js [hunk 12]
```diff
@@ -1,98 +1,343 @@
...
-    {
+  return url;
+};
+
+// Fetch access token from client id and client secret
+// DOC: https://developers.marketo.com/rest-api/authentication/
+const getAccessToken = async (config) =>
+  authCache.get(getAccessTokenCacheKey(config), async () => {
+    const url = getAccessTokenURL(config);
+    const { processedResponse: resp } = await handleHttpRequest('get', url, {
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply src/v0/destinations/marketo_bulk_upload/util.js:118]: done

File: src/v0/destinations/marketo_bulk_upload/util.js (hunk 7 above)
Comment (shrouti1507) [reply src/v0/destinations/marketo_bulk_upload/util.js:389]: done.

File: src/util/prometheus.js (hunk 1 above)
Comment (shrouti1507) [reply src/util/prometheus.js:889]: addressed

File: src/v0/destinations/marketo_bulk_upload/fileUpload.js [hunk 13]
```diff
@@ -168,32 +158,32 @@ const getFileData = async (inputEvents, config, fieldSchemaNames) => {
...
 
-const getImportID = async (input, config, fieldSchemaNames, accessToken) => {
+const getImportID = async (input, config, accessToken, csvHeader) => {
+  const importId = null; // by default importId is null
   const { readStream, successfulJobs, unsuccessfulJobs } = await getFileData(
     input,
     config,
-    fieldSchemaNames,
+    csvHeader,
   );
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply src/v0/destinations/marketo_bulk_upload/fileUpload.js:158]: addressed

File: src/v0/destinations/marketo_bulk_upload/poll.js [hunk 14]
```diff
@@ -18,106 +25,46 @@ const getPollStatus = async (event) => {
...
-        pollStatus.response.response.statusText || 'Error during polling status',
-        500,
-        pollStatus,
-      );
-    }
+
+  if (!isHttpStatusSuccess(pollStatus.status)) {
+    stats.counter(POLL_ACTIVITY, {
+      status: pollStatus.status,
+      state: 'Retryable',
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply src/v0/destinations/marketo_bulk_upload/poll.js:34]: addressed

**Review by chandumlg — DISMISSED** [review]

**Review by shrouti1507 — APPROVED** [review]

**Review by chandumlg — APPROVED** [review]

**Pull request merged**